### PR TITLE
Add OTLP exporter for Apollo metrics

### PR DESCRIPTION
**Pull request by BrynCooke**
This PR adds an OTLP metrics exporter for a future Apollo pipeline that can compliment the existing protobuf format.

It adds:
* many new metrics, although these won't be visible to users just yet.
* metrics around what was configured in the config file.
* filtering for existing and new metrics pipelines.

I want to get this merged before adding further metrics as the PR is getting large, and will become more difficult to review.

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/plugins/authentication/mod.rs [hunk 1]
```diff
@@ -403,6 +403,16 @@ fn authenticate(
             monotonic_counter.apollo_authentication_failure_count = 1u64,
             kind = %AUTHENTICATION_KIND
         );
+        tracing::info!(
+            monotonic_counter
+                .apollo
```
Comment (BrynCooke) [inline apollo-router/src/plugins/authentication/mod.rs:408]: metric name is wrong

Comment (BrynCooke) [reply apollo-router/src/plugins/authentication/mod.rs:474]: Fixed in [2def84a](https://github.com/apollographql/router/pull/3354/commits/2def84a927cc20ccec6fd0d9ba94d9b183676b78)

File: apollo-router/src/configuration/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    yaml: Value,
+    metrics: HashMap<String, (u64, HashMap<String, String>)>,
+}
+
+impl Metrics {
+    /// Spawn a task that will log configuration usage metrics every second.
+    /// This task has to run more frequently than that of the apollo otlp exporter otherwise the gauges will flap.
+    /// Dropping the MetricsHandle stops the task.  
+    pub(crate) async fn spawn(configuration: &Configuration) -> MetricsHandle {
+        let semaphore = Arc::new(tokio::sync::Semaphore::new(1));
```
Comment (o0Ignition0o) [inline apollo-router/src/configuration/metrics.rs:27]: nit, this is equivalent to using a oneshot channel

File: apollo-router/src/configuration/testdata/metrics/coprocessor.router.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        body: true
+        context: true
+        method: true
+      response:
+        status_code: true
+        headers: true
+        body: true
+        context: true
+        service_name: true
+
```
Comment (o0Ignition0o) [inline apollo-router/src/configuration/testdata/metrics/coprocessor.router.yaml:31]: TODO: update this or #3408 , whichever lands last.

File: apollo-router/src/context/mod.rs [hunk 4]
```diff
@@ -70,6 +70,13 @@ impl Context {
 }
 
 impl Context {
+    pub(crate) fn operation_name(&self) -> Option<String> {
+        // This method should be removed once we have a proper way to get the operation name.
```
Comment (o0Ignition0o) [inline apollo-router/src/context/mod.rs:74]: https://github.com/apollographql/router/blob/dd7bac0e9269fbe8c1b3be431cf9e4373862d084/apollo-router/src/services/layers/query_analysis.rs#L111 this looks to me like the right way to get the operation name (starting at the supergraph_service)

File: apollo-router/src/plugins/telemetry/metrics/filter.rs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        }
+    }
+
+    #[test]
+    fn test_apollo_metrics() {
+        let delegate = MockMeterProvider::default();
+        let filtered = FilterMeterProvider::apollo_metrics(delegate.clone())
+            .versioned_meter("filtered", None, None);
+        filtered.u64_counter("apollo.router.operations").init();
+        filtered.u64_counter("apollo.router.operation").init();
```
Comment (o0Ignition0o) [inline apollo-router/src/plugins/telemetry/metrics/filter.rs:218]: I don't see `apollo.router.operation` (singular) anywhere but in this test. Is this intentional?

File: apollo-router/src/services/layers/persisted_queries/mod.rs [hunk 6]
```diff
@@ -169,12 +170,17 @@ impl PersistedQueryLayer {
...
+                tracing::info!(monotonic_counter.apollo.router.operations.persisted_queries = 1u64);
                 Ok(request)
             } else if self.apq_enabled {
                 // if APQ is also enabled, pass the request along to the APQ plugin
                 // where it will do its own lookup
                 Ok(request)
             } else {
+                tracing::info!(
+                    monotonic_counter.apollo.router.operations.persisted_queries = 1u64,
+                    persisted_quieries.not_found = true
```
Comment (o0Ignition0o) [inline apollo-router/src/services/layers/persisted_queries/mod.rs:182]:
```suggestion
                    persisted_queries.not_found = true
```

File: apollo-router/src/plugins/telemetry/metrics/filter.rs (hunk 5 above)
Comment (o0Ignition0o) [reply apollo-router/src/plugins/telemetry/metrics/filter.rs:218]: removing it

File: apollo-router/src/plugins/telemetry/metrics/filter.rs [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        FilterMeterProvider::builder()
+            .delegate(delegate)
+            .deny(
+                Regex::new(r"apollo\.router\.(operations?|config)(\..*|$)")
+                    .expect("regex should have been valid"),
+            )
+            .allow(
+                Regex::new(r"apollo\.router\.operations\.authentication")
+                    .expect("regex should have been valid"),
+            )
```
Comment (Geal) [inline apollo-router/src/plugins/telemetry/metrics/filter.rs:56]: what's the rationale on what is sent to other exporters and what isn't? Example: I followed the new conventions for https://github.com/apollographql/router/commit/dd7bac0e9269fbe8c1b3be431cf9e4373862d084, but that new metric was first thought for usrs, to for analytics, so do I need to add an exemption here?

**Comment by BrynCooke** [comment]
> nit: that should have been multiple PRs. One for the OTLP layer, then other, smaller PRs for the metrics. It is hard to distinguish which changes are related. The changes related to the limits metrics in particular require looking at a lot of file to understand what's relevant

Agree. I've opened https://github.com/apollographql/router/pull/3595 which just contains the config metrics and exporter, but none of the operation level metrics.

**Comment by BrynCooke** [comment]
Closing as this PR was split into two smaller PRs.

**Pull request closed without merging**